packages/darklang/languageTools/lsp-server/fileSystemProvider.dark

```diff
@@ -0,0 +1,828 @@
+          | File
+          | Directory
+          | SymbolicLink
+
+        type FileSystemErrorType =
+          | FileNotFound
+          | FileExists
+          | FileNotADirectory
+          | FileIsADirectory
+          | NoPermissions
+          | Unavailable
+          | Other
+
+        type FileSystemProviderClientCapabilities = { queryable: Bool }
+
+        type URI = String
+
+        module FileSystemProviderOptions =
+          type FileSystemProviderOptions =
+            { scheme: String
+              caseSensitive: Bool
+              readonly: Bool }
+
+          let toJson (options: FileSystemProviderOptions) : Json =
+            [ Stdlib.Option.Option.Some(("scheme", Json.String options.scheme))
+              Stdlib.Option.Option.Some(
+                ("caseSensitive", Json.Bool options.caseSensitive)
+              )
+              Stdlib.Option.Option.Some(("readonly", Json.Bool options.readonly)) ]
+            |> Stdlib.Option.values
+            |> Json.Object
+
+
+        module ReadFile =
+          let method = "fileSystem/read"
+
+          module ReadFileParams =
+            type ReadFileParams = { uri: URI }
+
+            let fromJson
+              (json: Json)
+              : Stdlib.Result.Result<ReadFileParams, String> =
+              match json with
+              | Object fields ->
+                match
+                  Stdlib.List.findFirst fields (fun (key, _) -> key == "uri")
+                with
+                | Some((_, String uri)) ->
+                  (ReadFileParams { uri = uri }) |> Stdlib.Result.Result.Ok
+                | _ -> Stdlib.Result.Result.Error "Expected uri"
+              | _ -> Stdlib.Result.Result.Error "Expected object"
+
+
+          module ReadFileResponse =
+            type ReadFileResponse = { content: String }
+
+            let toJson (r: ReadFileResponse) : Json =
+              [ Stdlib.Option.Option.Some(("content", Json.String r.content)) ]
+              |> Stdlib.Option.values
+              |> Json.Object
+
+
+          let handleReadFileRequest
+            (state: LspState)
+            (requestId: JsonRPC.RequestId)
+            (params: ReadFileParams.ReadFileParams)
+            : LspState =
+            // Remove the darklang:// prefix and the .dark suffix from the URI, (e.g.darklang://type/Darklang/Stdlib/Option/Option.dark)
+            let uri =
+              params.uri |> Stdlib.String.dropFirst 11L |> Stdlib.String.dropLast 5L
+
+            let parts = uri |> Stdlib.String.split "/"
+
+            let elementType = parts |> Stdlib.List.head |> Builtin.unwrap
+
+            let nameForLookup =
+              parts |> Stdlib.List.tail |> Builtin.unwrap |> Stdlib.String.join "."
+
+            let pm = LanguageTools.PackageManager.pm ()
+            // e.g.type/Darklang/Stdlib/Option/Option
+            match elementType with
+            | "type" ->
+              let findType = pm.findType
+              let getType = pm.getType
```

Comment:
nitpick: don't think we need these helpers -- they're both only used once

Reply:
I believe these were added because we don't allow direct fieldAccess function calls (because of a parsing error)

Reply:
ah right -- OK, then maybe a CLEANUP to fix that when we have that interpreter/parsing stuff working.